lua/wire/zvm/zvm_tests.lua

```diff
@@ -0,0 +1,411 @@
+-- Library for making and running automated tests for Zyelios Virtual Machine Programs
+--
+--
+--------------------------------------------------------------------------------
+
+TESTING = true
+include("wire/cpulib.lua")
+include("wire/client/hlzasm/hc_compiler.lua")
+
+
+ZVMTestSuite = {
+	TestFiles = {},
+	TestQueue = {},
+	TestStatuses = {},
+	Warnings = 0,
+	CurrentWarnings = 0
+}
+
+local testDirectory = "wire/zvm/tests"
+
+function ZVMTestSuite.CMDRun(_,_,_,names)
+	ZVMTestSuite.Warnings = 0
+	ZVMTestSuite.TestFiles = {}
+	for filename in string.gmatch(names, '[^,]+') do
+		local files = file.Find('lua/'..testDirectory..'/'..filename..'.lua',"GAME")
+			for _,i in ipairs(files) do
+				ZVMTestSuite.TestFiles[#ZVMTestSuite.TestFiles+1] = i
+			end
+	end
+	if #ZVMTestSuite.TestFiles == 0 then
+		if names ~= nil then
+			if names ~= "" then
+				print("Didn't find any tests with name(s): "..names)
+				return
+			end
+		end
+		ZVMTestSuite.RunAll()
+	else
+		PrintTable(ZVMTestSuite.TestFiles)
+		ZVMTestSuite.StartTesting()
+	end
+end
+
+function ZVMTestSuite.RunAll()
+	local files,directories = file.Find(testDirectory..'/*.lua',"LUA","nameasc")
+	ZVMTestSuite.TestFiles = files or {}
+	ZVMTestSuite.StartTesting()
+end
+
+function ZVMTestSuite.StartTesting()
+	ZVMTestSuite.TestQueue = {}
+	ZVMTestSuite.TestStatuses = {}
+	for ind,i in ipairs(ZVMTestSuite.TestFiles) do -- copy with reversed indexes so we can use cheap popping
+		ZVMTestSuite.TestQueue[(#ZVMTestSuite.TestFiles)+1-ind] = i
+	end
+	print(#ZVMTestSuite.TestFiles.." tests loaded")
+	ZVMTestSuite.RunNextTest()
+end
+
+function ZVMTestSuite.FinishTest(fail)
+	local finalFail = false
+	if fail == nil then
+		finalFail = true
+	else
+		finalFail = fail
+	end
+	if ZVMTestSuite.CurrentWarnings > 0 then
+		print("Compiler Warnings from "..ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue]..": "..ZVMTestSuite.CurrentWarnings)
+		ZVMTestSuite.CurrentWarnings = 0
+	end
+	ZVMTestSuite.TestStatuses[#ZVMTestSuite.TestStatuses+1] = finalFail -- auto fail on return nil
+	ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue] = nil
+	if #ZVMTestSuite.TestQueue > 0 then
+		ZVMTestSuite.RunNextTest()
+	else
+		local passed, failed = 0, 0
+		for ind,i in ipairs(ZVMTestSuite.TestFiles) do
+			if ZVMTestSuite.TestStatuses[ind] then
+				failed = failed + 1
+				MsgC(Color(255,0,0),"Error ",Color(255,255,255),"in "..i.."\n")
+			else
+				passed = passed + 1
+				--MsgC(Color(0,255,0),i.." passed tests".."\n")
+			end
+		end
+		local passmod, errormod = "",""
+		if passed > 1 then
+			passmod = "s"
+		end
+		if failed > 1 then
+			errormod = "s"
+		end
+		print(failed.." Failed test"..errormod..", "..passed.." Passed test"..passmod..", Compiler Warnings: "..ZVMTestSuite.Warnings)
+	end
+end
+
+function ZVMTestSuite.Error(...)
+	local args = table.Pack(...)
+	MsgC(Color(255,0,0),"in file ",Color(255,255,255),ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue],Color(255,0,0)," Error: ")
+	if args ~= nil then
```

Comment:
This whole part could be simplified to just `MsgC(color_white, ...)` no?

Feel like this is needlessly complex. Also don't really like the Color() spam in here, even if it is just for debugging, at least could extract to constant variables